Solutions/Corvus.Azure.Storage.Tenancy/Corvus/Azure/Storage/Tenancy/Internal/TenantCloudTableFactory.cs

```diff
@@ -0,0 +1,303 @@
+            {
+                throw new System.ArgumentNullException(nameof(storageConfiguration));
+            }
+
+            return string.IsNullOrEmpty(storageConfiguration.AccountName) ? "storageConfiguration-developmentStorage" : $"storageConfiguration-{storageConfiguration.AccountName}";
+        }
+
+        /// <summary>
+        /// Get a table for a tenant.
+        /// </summary>
+        /// <param name="tenant">The tenant for which to retrieve the table.</param>
+        /// <param name="tableDefinition">The details of the table to create.</param>
+        /// <returns>The container instance for the tenant.</returns>
+        /// <remarks>
+        /// This caches table instances to ensure that a singleton is used for all request for the same tenant and table definition.
+        /// </remarks>
+        public Task<CloudTable> GetTableForTenantAsync(ITenant tenant, TableStorageTableDefinition tableDefinition)
+        {
+            if (tenant is null)
+            {
+                throw new System.ArgumentNullException(nameof(tenant));
+            }
+
+            if (tableDefinition is null)
+            {
+                throw new System.ArgumentNullException(nameof(tableDefinition));
+            }
+
+            TableStorageTableDefinition tenantedTableStorageTableDefinition = BuildTableStorageTableDefinitionForTenant(tenant, tableDefinition);
+            object key = GetKeyFor(tenantedTableStorageTableDefinition);
+
+            return this.containers.GetOrAdd(
+                key,
+                async _ => await this.CreateTenantCloudTable(tenant, tableDefinition, tenantedTableStorageTableDefinition).ConfigureAwait(false));
+        }
+
+        /// <summary>
+        /// Create the table instance.
+        /// </summary>
+        /// <param name="tenant">The tenant.</param>
+        /// <param name="tenantedTableStorageTableDefinition">The table definition, adapted for the tenant.</param>
+        /// <param name="configuration">The table configuration.</param>
+        /// <returns>A <see cref="Task"/> which completes with the instance of the table for the tenant.</returns>
+        protected async Task<CloudTable> CreateCloudTableInstanceAsync(
+            ITenant tenant,
+            TableStorageTableDefinition tenantedTableStorageTableDefinition,
+            TableStorageConfiguration configuration)
+        {
+            if (tenant is null)
+            {
+                throw new System.ArgumentNullException(nameof(tenant));
+            }
+
+            if (tenantedTableStorageTableDefinition is null)
+            {
+                throw new System.ArgumentNullException(nameof(tenantedTableStorageTableDefinition));
+            }
+
+            if (configuration is null)
+            {
+                throw new System.ArgumentNullException(nameof(configuration));
+            }
+
+            // Null forgiving operator only necessary for as long as we target .NET Standard 2.0.
```

Comment:
I have realised fairly recently that if we multi-target, we can configure the project so that we enable annotations for both .NET Standard 2.0 and later, but that we only enable warnings for standard2.1 or core3.1 and later. This means you stop seeing the spurious warnings, and don't need to litter the code with `!`.

Reply:
I will leave you to add the relevant issues for this change :)